src/analytics-pre-init.ts

```diff
@@ -0,0 +1,276 @@
+import { Analytics } from './analytics'
+import { Context } from './core/context'
+import { allSettled } from './lib/all-settled'
+
+/**
+ * The names of any Analytics instance methods that can be called pre-initialization.
+ * These methods should exist statically on AnalyticsBrowser.
+ */
+type PreInitMethodName =
+  | 'trackSubmit'
+  | 'trackClick'
+  | 'trackLink'
+  | 'trackForm'
+  | 'pageview'
+  | 'identify'
+  | 'reset'
+  | 'group'
+  | 'track'
+  | 'ready'
+  | 'alias'
+  | 'debug'
+  | 'page'
+  | 'once'
+  | 'off'
+  | 'on'
+  | 'addSourceMiddleware'
+  | 'addIntegrationMiddleware'
+  | 'setAnonymousId'
+  | 'addDestinationMiddleware'
+
+function getMethodCallsByMethodName<
+  T extends PreInitMethodName,
+  U extends PreInitMethodCall[]
+>(methodNames: T[], calls: U): PreInitMethodCall<T>[] {
+  return calls.filter((call): call is PreInitMethodCall<T> =>
+    methodNames.includes(call.method as T)
+  )
+}
+
+export function flushAnalyticsCallsInNewTask(
+  analytics: Analytics,
+  calls: PreInitMethodCall[]
+): void {
+  calls.forEach((m) => {
+    setTimeout(() => {
+      callAnalyticsMethod(analytics, m).catch(console.error)
+    }, 0)
+  })
+}
+
+async function flushAnalyticsCallsConcurrentlyByName(
+  name: PreInitMethodName,
+  analytics: Analytics,
+  calls: PreInitMethodCall[]
+): Promise<void> {
+  await allSettled(
+    getMethodCallsByMethodName([name], calls).map((c) => {
+      return callAnalyticsMethod(analytics, c).catch(console.error)
+    })
+  )
+}
+
+async function flushAnalyticsCallsSeriallyByName(
+  name: PreInitMethodName,
+  analytics: Analytics,
+  calls: PreInitMethodCall[]
+): Promise<void> {
+  for (const c of calls) {
+    if (c.method === name) {
+      await callAnalyticsMethod(analytics, c).catch(console.error)
+    }
+  }
+}
+
+export const flushAddSourceMiddleware = flushAnalyticsCallsSeriallyByName.bind(
```

Comment:
The previous logic had multiple methods of "addSourceMiddleware" being called serially, rather than concurrently -- so I created a function to handle this case. This potentially slows down full initialization (however that matters); However, there were no tests asserting that -- so I am still puzzling together the previous intention. It's pretty common for devs to be confused in javascript about the async behavior in loops vs higher-order-functions vs Promise.whatever, so I never rule that out. 

So my outstanding questions is:
-  can addSourceMiddleware be safely refactored to be called concurrently using promise.allSettled, like the other methods?

Reply:
That's a good point about the order of middleware.

The "concurrent" function that uses `allSettled` would still preserve the order that the middleware is registered, since the Array.map/filter calls that creates the promises will operate on the middleware functions in order.

The only possibilities I could think of:
 - it's a mistake -- serial calls are not necessary (for example, promise.allSettled would work)
 - there's a race condition around registering middleware too quickly, but it's not caught in the tests.

Reply:
We can add some emit functionality to curtail this issue where we emit a "done" message once all the middlewares are finished registering, and then wait on that for the rest of the tasks.